---
title: "Working with STAR outputs"
highlight: "tango"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette demonstrates how to read, modify, and merge output files generated by the STAR RNA-seq aligner using functions from the `ngsmisc` package.  
We will cover two main types of STAR outputs: `final.log` files, which summarize alignment statistics, and `SJ.out.tab` files, which contain splice junction information.  
The examples below show how to process these files, combine results from multiple samples, and visualize the data.

# Reading, Modifying, and Merging `final.log` Files

The `final.log` file produced by STAR contains summary statistics for each alignment run.  
The `ngsmisc` package provides functions to read/merge these files.

```{r}
# Locate example STAR final.log output files included in the ngsmisc package.
infs <-
  system.file(package = "ngsmisc", "star") |>
  fs::dir_ls(regexp = "final.log$")

# Display the contents of the first final.log file for inspection.
infs[1] |> readLines() |> cat(sep = "\n")

# Read a final.log file using ST_read_final_log().
# The result is a tibble with parsed statistics.
ngsmisc::ST_read_final_log(infs[1]) |>
  ngsmisc::rename_fpath(nth = 3)

# Read and merge multiple final.log files.
# Each file is read and renamed for clarity, then merged into a single tibble.
infs |>
  lapply(ngsmisc::ST_read_final_log) |>
  lapply(ngsmisc::rename_fpath, nth = 3, suffix = ".final.log") |>
  ngsmisc::ST_merge_final_log()
```

# Reading, Modifying, and Merging `SJ.out.tab` Files

The `SJ.out.tab` file contains information about detected splice junctions.
`ngsmisc` provides functions to read these files, decode categorical columns.

```{r}
# Locate example STAR SJ.out.tab output files included in the ngsmisc package.
infs <-
  system.file(package = "ngsmisc", "star") |>
  fs::dir_ls(regexp = ".sj.tsv$")

# Display the first 10 lines of the first SJ.out.tab file.
infs[1] |> readLines(n = 10) |> cat(sep = "\n")

# Read an SJ.out.tab file using ST_read_sj_tab().
ngsmisc::ST_read_sj_tab(infs[1])

# Decode integer columns (strand, intron_motif, annotated) to human-readable strings.
ngsmisc::ST_read_sj_tab(infs[1], decode = TRUE)

# Convert the decoded table to a GenomicRanges::GRanges object for genomic analysis.
ngsmisc::ST_read_sj_tab(infs[1], decode = TRUE) |> plyranges::as_granges()

# Merge multiple SJ.out.tab files, add sample information, and combine into a single tibble.
tbl_merge <-
  infs |>
  lapply(ngsmisc::ST_read_sj_tab, decode = TRUE) |>
  lapply(head) |>
  purrr::imap(~ dplyr::mutate(.x, sample = ngsmisc::renamer_fpath()(.y))) |>
  dplyr::bind_rows()
tbl_merge

# Visualize the number of uniquely mapped reads crossing each junction for each sample.
library(ggplot2)
tbl_merge |>
  ggplot(aes(x = paste(seqnames, start, end, sep = "\n"), y = num_uniq_map_jc)) +
  geom_col(aes(fill = sample)) +
  theme_linedraw(base_size = 14) +
  facet_grid(cols = vars(intron_motif), scales = "free_x",  space = "free") +
  labs(
    x = "Coordinates",
    y = "Number of uniquely mapped reads\ncrossing junction"
  ) +
  scale_y_continuous(limits = c(0, NA), expand = expansion(c(0, .1)))
```

# Sessioninfo
```{r}
sessioninfo::package_info()
```
